Add WebSocket compression support

BrennanConroy opened the pull request:
Fixes https://github.com/dotnet/aspnetcore/issues/2715

Main questions are around usability. Do we want to allow setting specific deflate options or just have a bool for enabling compression and let the client negotiate choose what it wants to use?

Added an `AcceptWebSocketAsync` overload to `WebSocketManager` to see how it looks to pass per connection options.

https://tools.ietf.org/html/rfc7692#section-5 for general negotiation details
https://tools.ietf.org/html/rfc7692#section-7 for permessage-deflate details

```diff
namespace Microsoft.Net.Http.Headers
{
    public static class HeaderNames
    {
+        public static readonly string SecWebSocketExtensions = "Sec-WebSocket-Extensions";
    }
}

namespace Microsoft.AspNetCore.Http
{
    public abstract class WebSocketManager
    {
+        public virtual Task<WebSocket> AcceptWebSocketAsync(WebSocketAcceptContext acceptContext) => throw new NotImplementedException();
    }
}

namespace Microsoft.AspNetCore.Http
{
    public partial interface IHeaderDictionary
    {
+        StringValues SecWebSocketExtensions { get; set; }
    }
}

namespace Microsoft.AspNetCore.Http
{
    internal sealed class DefaultWebSocketManager : WebSocketManager
    {
+        public override Task<WebSocket> AcceptWebSocketAsync(WebSocketAcceptContext acceptContext);
    }
}

namespace Microsoft.AspNetCore.WebSockets
{
    public class ExtendedWebSocketAcceptContext : WebSocketAcceptContext
    {
+        public bool DangerousEnableCompression { get; set; }
+        public bool DisableServerContextTakeover { get; set; } = false;
+        public int ServerMaxWindowBits { get; set; } = 15;
    }
}

namespace Microsoft.AspNetCore.Http
{
+    public static class WebSocketManagerExtensions
+    {
+        public static Task<WebSocket> AcceptWebSocketAsync(this WebSocketManager webSocketManager, ExtendedWebSocketAcceptContext acceptContext);
+    }
}
```

BrennanConroy committed:
Initial WebSocket compression implementation

BrennanConroy committed:
fix and some tests

BrennanConroy committed:
More compression

Tratcher commented on src/Middleware/WebSockets/src/HandshakeHelpers.cs:
```diff
@@ -72,5 +75,89 @@ public static string CreateResponseKey(string requestKey)
+                {
+                    if (value.SequenceEqual(WebSocketDeflateConstants.ClientNoContextTakeover))
+                    {
+                        // REVIEW: If someone specifies true for server options, do we allow client to override this?
```
> By including this extension parameter in an
   extension negotiation offer, a client informs the peer server of a
   hint that even if the server doesn't include the
   "client_no_context_takeover" extension parameter in the corresponding
   extension negotiation response to the offer, the client is not going
   to use context takeover.

It doesn't seem to matter if the client 'overrides' the server, the client isn't going to use takeover regardless.

Tratcher commented on src/Middleware/WebSockets/src/HandshakeHelpers.cs:
```diff
@@ -72,5 +75,89 @@ public static string CreateResponseKey(string requestKey)
+                    else if (value.SequenceEqual(WebSocketDeflateConstants.ServerNoContextTakeover))
+                    {
+                        options.ServerContextTakeover = false;
+                        builder.Append("; ").Append(WebSocketDeflateConstants.ServerNoContextTakeover);
```
>   A server MAY include the "server_no_context_takeover" extension
   parameter in an extension negotiation response even if the extension
   negotiation offer being accepted by the extension negotiation
   response didn't include the "server_no_context_takeover" extension
   parameter.

Should we add it to the response if ServerContextTakeover = false was set by the server rather than the client?

Tratcher commented on src/Middleware/WebSockets/src/HandshakeHelpers.cs:
```diff
@@ -72,5 +75,89 @@ public static string CreateResponseKey(string requestKey)
             return Convert.ToBase64String(hashedBytes);
         }
+
+        public static bool ParseDeflateOptions(ReadOnlySpan<char> extension, WebSocketDeflateOptions options, [NotNullWhen(true)] out string? response)
```
```suggestion
        // https://datatracker.ietf.org/doc/html/rfc7692#section-7.1
        public static bool ParseDeflateOptions(ReadOnlySpan<char> extension, WebSocketDeflateOptions options, [NotNullWhen(true)] out string? response)
```

Tratcher commented on src/Middleware/WebSockets/src/HandshakeHelpers.cs:
```diff
@@ -72,5 +75,89 @@ public static string CreateResponseKey(string requestKey)
+                        }
+
+                        if (!int.TryParse(value[(startIndex + 1)..], NumberStyles.Integer, CultureInfo.InvariantCulture, out int windowBits) ||
+                            windowBits < 9 ||
```
The spec says 8 is the lower bound. There was an issue with the implementation not supporting that, right? Even so, it shouldn't generate an exception here, it should be politely ignored (ParseDeflateOptions returns false). Include a comment that 8 is a valid value but unsupported.

Tratcher commented on src/Middleware/WebSockets/src/HandshakeHelpers.cs:
```diff
@@ -72,5 +75,89 @@ public static string CreateResponseKey(string requestKey)
+                        builder.Append("; ").Append(WebSocketDeflateConstants.ClientMaxWindowBits).Append('=')
+                            .Append(options.ClientMaxWindowBits.ToString(CultureInfo.InvariantCulture));
+                    }
+                    else if (value.StartsWith(WebSocketDeflateConstants.ServerMaxWindowBits))
```
The server still needs to send back these limits even if it doesn't receive it from the client.

BrennanConroy commented on src/Middleware/WebSockets/src/WebSocketMiddleware.cs:
```diff
@@ -139,21 +144,117 @@ public async Task<WebSocket> AcceptAsync(WebSocketAcceptContext acceptContext)
                 HandshakeHelpers.GenerateResponseHeaders(key, subProtocol, _context.Response.Headers);
 
+                WebSocketDeflateOptions? deflateOptions = null;
+                var ext = _context.Request.Headers.SecWebSocketExtensions;
```
Whoops 😄

BrennanConroy committed:
getting closer

BrennanConroy commented:
* Changed the options to only have server side settings
  - We could make the settings nullable and if either is set the assume compression is enabled (so we can remove the bool)
* Question: do we reject server_no_context if servercontext=true?

Tratcher commented:
> * We could make the settings nullable and if either is set then assume compression is enabled (so we can remove the bool)

No, better to have an explicit on-off switch.

Tratcher commented:
> * Question: do we reject server_no_context if servercontext=true?

> If the peer server doesn't use context
   takeover, the client doesn't need to reserve memory to retain the
   LZ77 sliding window between messages.

Sounds like the client is supposed to be able to opt out, so it's fine for them to override servercontext=true.

Tratcher commented on src/Middleware/WebSockets/src/WebSocketMiddleware.cs:
```diff
@@ -139,21 +144,77 @@ public async Task<WebSocket> AcceptAsync(WebSocketAcceptContext acceptContext)
+                        {
+                            if (extension.TrimStart().StartsWith("permessage-deflate", StringComparison.Ordinal))
+                            {
+                                deflateOptions = new WebSocketDeflateOptions()
+                                {
+                                    ServerContextTakeover = serverContextTakeover,
+                                    ServerMaxWindowBits = serverMaxWindowBits
+                                };
+                                if (HandshakeHelpers.ParseDeflateOptions(extension, deflateOptions, out var response))
```
deflateOptions is still set even if ParseDeflateOptions returns false, which would result in the server using compression but the client not knowing it.

Change deflateOptions to an out paramter and pass in the serverContextTakeover and serverMaxWindowBits so deflateOptions only gets set if ParseDeflateOptions returns true.

Tratcher commented on src/Middleware/WebSockets/src/WebSocketMiddleware.cs:
```diff
@@ -139,21 +144,77 @@ public async Task<WebSocket> AcceptAsync(WebSocketAcceptContext acceptContext)
+                });
+            }
+
+            private static WebSocketDeflateOptions? CloneWebSocketDeflateOptions([NotNullIfNotNull("options")] WebSocketDeflateOptions? options)
```
No longer needed.

Tratcher commented on src/Middleware/WebSockets/test/UnitTests/WebSocketCompressionMiddlewareTests.cs:
```diff
@@ -0,0 +1,306 @@
+
+namespace Microsoft.AspNetCore.WebSockets.Test
+{
+    public class WebSocketCompressionMiddlewareTests : LoggedTest
```
Have you enabled the autobahn tests yet?

BrennanConroy replied:
I ran it earlier, but wanting to get everything cleaned up before trying again

BrennanConroy replied:
All Autobahn tests passed

BrennanConroy commented:
> Sounds like the client is supposed to be able to opt out, so it's fine for them to override servercontext=true.

At that point, why even expose an option for this? Just in the case where the client doesn't specify an option?

Tratcher reviewed (commented):
> At that point, why even expose an option for this? Just in the case where the client doesn't specify an option?

So the server can also opt-out. It sounds like context takeover uses more memory on both client and server.

Tratcher commented on src/Middleware/WebSockets/src/HandshakeHelpers.cs:
```diff
@@ -72,5 +75,136 @@ public static string CreateResponseKey(string requestKey)
+                int end = extension.IndexOf(';');
+                ReadOnlySpan<char> value = (end >= 0 ? extension[..end] : extension).Trim();
+
+                if (value.Length > 0)
+                {
```
Style: you should be able to reduce the nesting here

Tratcher commented on src/Middleware/WebSockets/src/HandshakeHelpers.cs:
```diff
@@ -72,5 +75,136 @@ public static string CreateResponseKey(string requestKey)
+                            windowBits < 8 ||
+                            windowBits > 15)
+                        {
+                            throw new WebSocketException(WebSocketError.HeaderError, $"invalid {propertyName} used: {value[(startIndex + 1)..].ToString()}");
```
Exceptions caused by external user data... Probably better to ignore it (maybe log it), and return false.

BrennanConroy committed:
update

BrennanConroy commented:
:up: :date:
added support for quoted values, duplicate parameters from a negotiate request is invalid so added that, renamed the server context takeover option to make it obvious that it's an opt-out, changed the base parsing tests to not use a server.

BrennanConroy committed:
generated

Tratcher commented on src/Middleware/WebSockets/src/WebSocketMiddleware.cs:
```diff
@@ -139,21 +144,57 @@ public async Task<WebSocket> AcceptAsync(WebSocketAcceptContext acceptContext)
+                        {
+                            if (extension.AsSpan().TrimStart().StartsWith("permessage-deflate", StringComparison.Ordinal))
+                            {
+                                if (HandshakeHelpers.ParseDeflateOptions(extension.AsSpan().TrimStart(), serverContextTakeover, serverMaxWindowBits, out var parsedOptions, out var response))
```
We'll want some debug logging around the success or failure of this negotiation and the resulting values that will be used.

Tratcher commented on src/Middleware/WebSockets/src/HandshakeHelpers.cs:
```diff
@@ -72,5 +75,186 @@ public static string CreateResponseKey(string requestKey)
+
+                    // https://datatracker.ietf.org/doc/html/rfc7692#section-5.2
+                    // check for value in quotes and pull the value out without the quotes
+                    if (value[0] == '"' && value.EndsWith("\"".AsSpan()))
+                    {
+                        value = value[1..^1];
+                    }
```
Does this handle the error case of a single quote `foo="`?

BrennanConroy replied:
Without a number? Will add a test for that

Tratcher commented on src/Middleware/WebSockets/src/ExtendedWebSocketAcceptContext.cs:
```diff
@@ -24,5 +25,20 @@ public class ExtendedWebSocketAcceptContext : WebSocketAcceptContext
         public TimeSpan? KeepAliveInterval { get; set; }
+
+        /// <summary>
+        ///
```
Docs

Tratcher reviewed (commented):
It's looking really good. What's left, autobahn?

BrennanConroy commented:
>What's left, autobahn?

Yeah, need to do another round of that, would like to figure out if we can make the `AcceptWebSocketAsync` api with the `ExtendedWebSocketAcceptContext` discoverable.

BrennanConroy committed:
fb

BrennanConroy commented:
Would we be against moving the new properties to the base WebSocketAcceptContext?

Tratcher commented:
> Would we be against moving the new properties to the base WebSocketAcceptContext?

We'd originally left them separate because the base WebSocketAcceptContext was supposed to be implementation agnostic. Now that we only use one implementation I'm not sure if that matters anymore.

> Or maybe we can make an extension in the websocket project for `AcceptWebSocketAsync(ExtendedWebSocketAcceptContext)`

Let's start with that.

BrennanConroy committed:
extension

BrennanConroy commented:
The extension method is a bit weird, it doesn't get called when passing an `ExtendedWebSocketAcceptContext`, but it at least shows in intellisense.

BrennanConroy commented:
API review:
Remove AcceptWebSocketAsync extension
Move options from ExtendedWebSocketAcceptContext to WebSocketAcceptContext
Make KeepAliveInterval virtual on WebSocketAcceptContext
Obsolete ExtendedWebSocketAcceptContext

davidfowl commented:
> Make KeepAliveInterval virtual on WebSocketAcceptContext

Turns out this is a breaking change.

Tratcher commented:
So how do we consolidate ExtendedWebSocketAcceptContext?

BrennanConroy commented:
>❌ DISALLOWED: Adding the virtual keyword to a member

We're adding a member to `WebSocketAcceptContext`, not making an existing member virtual so does it actually fall under this category?

We're still going to `as` cast to `ExtendedWebSocketAcceptContext` in the middleware, so old usage should still work, and new code will use the base type and we'll reference the new member.

BrennanConroy committed:
api review

BrennanConroy committed:
fix

davidfowl commented on src/Http/Http/src/Internal/DefaultWebSocketManager.cs:
```diff
@@ -61,12 +61,17 @@ public override IList<string> WebSocketRequestedProtocols
 
         public override Task<WebSocket> AcceptWebSocketAsync(string? subProtocol)
+        {
+            return AcceptWebSocketAsync(new WebSocketAcceptContext() { SubProtocol = subProtocol });
```
Allocations.... singleton when subprotocol is null?

davidfowl commented on src/Middleware/WebSockets/src/HandshakeHelpers.cs:
```diff
@@ -72,5 +75,89 @@ public static string CreateResponseKey(string requestKey)
+        public static bool ParseDeflateOptions(ReadOnlySpan<char> extension, WebSocketDeflateOptions options, [NotNullWhen(true)] out string? response)
+        {
+            response = null;
+            var builder = new StringBuilder(WebSocketDeflateConstants.MaxExtensionLength);
```
Performance 😢

BrennanConroy replied:
Ok, want me to add the logic to calculate the length then use string.Create to populate the span?

BrennanConroy replied:
Crazy code added

Tratcher commented on src/Http/Http.Features/src/WebSocketAcceptContext.cs:
```diff
@@ -14,5 +15,42 @@ public class WebSocketAcceptContext
+
+        /// <summary>
+        /// Enables support for the 'permessage-deflate' WebSocket extension.<para />
+        /// Be aware that enabling compression makes the application subject to CRIME/BREACH type attacks.
```
```suggestion
        /// Be aware that enabling compression over encrypted connections makes the application subject to CRIME/BREACH type attacks.
```

Tratcher commented on src/Http/Http.Features/src/WebSocketAcceptContext.cs:
```diff
@@ -14,5 +15,42 @@ public class WebSocketAcceptContext
+        public bool DangerousEnableCompression { get; set; }
+
+        /// <summary>
+        /// Disables server context takeover when using compression.
```
This can be used to lower memory utilization, right? Please include that 'why'.

BrennanConroy replied:
Done

Tratcher commented on src/Http/Http.Features/src/WebSocketAcceptContext.cs:
```diff
@@ -14,5 +15,42 @@ public class WebSocketAcceptContext
+        public bool DisableServerContextTakeover { get; set; }
+
+        /// <summary>
+        /// Sets the maximum base-2 logarithm of the LZ77 sliding window size that can be used for compression.
```
Why would I set this property?

What is the valid range?

There should be argument validation on the setter.

BrennanConroy replied:
Also done

BrennanConroy committed:
fb

davidfowl commented on src/Http/Http/src/Internal/DefaultWebSocketManager.cs:
```diff
@@ -62,7 +63,9 @@ public override IList<string> WebSocketRequestedProtocols
 
         public override Task<WebSocket> AcceptWebSocketAsync(string? subProtocol)
         {
-            return AcceptWebSocketAsync(new WebSocketAcceptContext() { SubProtocol = subProtocol });
```
👍🏾

davidfowl commented on src/Middleware/WebSockets/src/HandshakeHelpers.cs:
```diff
@@ -248,17 +246,69 @@ static bool ParseWindowBits(ReadOnlySpan<char> value, string propertyName, out i
             }
 
-            response = builder.ToString();
+            response = string.Create(responseLength, (parsedOptions, hasClientMaxWindowBits, hasServerMaxWindowBits, hasClientNoContext),
```
Another option would be to use https://github.com/dotnet/runtime/blob/5525d0bbb57d5550d71a0a15158f992b0e694000/src/libraries/Common/src/System/Text/ValueStringBuilder.cs

BrennanConroy replied:
Done, looks a bit nicer 😃

BrennanConroy committed:
wip

BrennanConroy committed:
valuestringbuilder

BrennanConroy committed:
copied from
